### ISSUE TITLE
2FA PIN Reset

### ISSUE THREAD

**Issue by rachaelwoolen** [seedco/Workshop#2]
### What
Reset PIN via the app using 2FA verification.

### Description
If and when a customer needs to reset their PIN, they should be able to do so via the app and without needing to contact support. 

1. Login
2. Select Reset PIN
3. We text a code to customer's phone number on file
4. Customer enters code 
5. Customer chooses new 4 digit PIN and verifies it

2FA means the customer doesn't have to contact support to reset their PIN, thus reducing contact and minimizing the risk of fraud. It also means the customer can securely reset their PIN without having to remember the old one. 

### User Story
As a Seed customer, I probably need to reset my PIN for one of two reasons. Either I think my PIN has been compromised or I have forgotten the PIN I selected. If the latter is true, I would not currently be able to reset my PIN without contacting support. I want to be able to verify my identity and reset my PIN securely without having to remember the old PIN or contacting support.

**Comment by benjisg** [comment]
Related slack [discussion](https://seedinc.slack.com/archives/C594H6S79/p1494888008635956)

The user story here is really covering two different scenarios; a pin change needed when they remember their pin and one needed when they don't. I'm going to take this from the point of the second one since we already support in-app pin changes if they know their current pin (ignoring PIN attempt lockout issues).

The best route for this one would likely to be implementing something like this:

1. Request pin reset from Support
2. Support confirms identity of requestor
3. Support requests 2FA code sent to customer SMS number
4. Customer can use 2FA code in app to unlock 1 use of pin reset with no existing pin validation

This flow has the following benefits:
* Pin resets are an exception activity and we should confirm the ID of the person requesting a reset 
* Seed employees (dev or support) will never need to know a temporary pin for the individual
* Pins would not need to be transmitted through email, SMS, or a support person. The app has security measures in place to ensure that requests are talking to our servers, making it a closed loop
* 2FA codes can be set to expire if not used within a given time period
* Pin change would still require access to the app (Touch ID or passcode) and reduces impact surface area for social engineering of support

**Comment by rachaelwoolen** [comment]
@benjisg If I understand correctly, what your proposing is something similar if the customer reports they can't remember their PIN, but would require to them to call support?

If we are able to access TBBK's api to reset PIN's after too many failed attempts, would that flow also apply?

**Comment by benjisg** [comment]
Yeah so it's more of a question of how comfortable we are allowing someone to reset their pin without knowing their previous one while bypassing any secondary checks. Come to think of it we could likely argue that Touch ID/app passcode is providing our secondary form of authentication to allow them to trigger that option from within the app and bypass talking to support all together. The chances that someone steals a debit card and can get access to the reset functionality through the app are significantly low.

As for auto-resetting to prevent lockouts that could be tough currently because we only refresh txn data every 10 minutes and they could do several failed pin attempts at an ATM or terminal during that time. I think even if we were pulling data every minute or so it'd likely still lead to a lot of resets. We probably want to focus on getting TBBK to raise the threshold of how many failures cause a lockout or (better yet) give us a way to unlock the pin ourselves.

**Comment by rachaelwoolen** [comment]
I agree that having touch ID/passcode security on the app should be enough. I just think that asking customer's for their old PIN when trying to change it is unrealistic and could lead to more contact down the road. 

> We probably want to focus on getting TBBK to raise the threshold of how many failures cause a lockout or (better yet) give us a way to unlock the pin ourselves.

Either, or better yet, both of those sounds good.
